### PR TITLE
Adding support for Spies in Code Generation

### PR DESCRIPTION
**Pull request by MarcoSignoretto**
This patch allows the new ksp based code generator to generate Spies in addition to mocks

TODO:
- [x] Continue properties implementation
- [x] Compare generated code for mocks before and after
- [x] Test Spy generation on an internal project
- [x] Update README

### CODE REVIEW

File: mockingbird-processor/src/main/kotlin/com/careem/mockingbird/processor/CodeBlockFactory.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeSpec
+
+interface CodeBlockFactory {
+    fun resolveSupertype(): String
+    fun decorateConstructor(classToMock: KSType, classBuilder: TypeSpec.Builder)
+    fun decorateFunctionBody(classToMock: KSClassDeclaration, function: KSFunctionDeclaration, isUnit: Boolean, functionBuilder: FunSpec.Builder)
+    fun decoratePropertyGetter(classToMock: KSClassDeclaration, property: KSPropertyDeclaration, propertyBuilder: PropertySpec.Builder, typeResolver: Map<KSTypeParameter, KSTypeArgument>)
+    fun decoratePropertySetter(classToMock: KSClassDeclaration, property: KSPropertyDeclaration, propertyBuilder: PropertySpec.Builder, typeResolver: Map<KSTypeParameter, KSTypeArgument>)
+}
```
Comment (MarcoSignoretto) [inline mockingbird-processor/src/main/kotlin/com/careem/mockingbird/processor/CodeBlockFactory.kt:35]: this si where codegen is different in Spies respect to mocks

File: mockingbird-processor/src/main/kotlin/com/careem/mockingbird/processor/Generator.kt [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+import com.squareup.kotlinpoet.MemberName
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.buildCodeBlock
+import com.squareup.kotlinpoet.ksp.toClassName
+import com.squareup.kotlinpoet.ksp.toKModifier
+import com.squareup.kotlinpoet.ksp.toTypeName
+
+abstract class Generator constructor(
```
Comment (MarcoSignoretto) [inline mockingbird-processor/src/main/kotlin/com/careem/mockingbird/processor/Generator.kt:40]: just rename of MockGenerator, non common logic has been extracted to `MockGenerator` and `SpyGenerator`

File: mockingbird-processor/src/main/kotlin/com/careem/mockingbird/processor/SpyGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                    .build()
+            )
+            .addProperty(
+                PropertySpec.builder(propertyName, className)
+                    .initializer(propertyName)
+                    .build()
+            )
+    }
+
+    override fun decorateFunctionBody(
```
Comment (MarcoSignoretto) [inline mockingbird-processor/src/main/kotlin/com/careem/mockingbird/processor/SpyGenerator.kt:58]: function body is different between mocks and spies, this is the spy impl for the code gen of that block

File: samples/kspsample/src/commonTest/kotlin/com/careem/mockingbird/kspsample/KspSampleSpyTest.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import com.careem.mockingbird.test.annotations.Spy
+import kotlin.test.Test
+import kotlin.test.assertNotNull
+import kotlin.test.assertNull
+
+class KspSampleSpyTest {
+
+//    @Spy
+//    lateinit var pippoMock: PippoSample
+// FIXME suspend function is not working
```
Comment (MarcoSignoretto) [inline samples/kspsample/src/commonTest/kotlin/com/careem/mockingbird/kspsample/KspSampleSpyTest.kt:28]: this will be supported in future PRs

**Review by carlonzo — APPROVED** [review]

**Pull request merged**